cmd/gaia/init/testnet.go

```diff
@@ -182,92 +200,148 @@ func testnetWithConfig(config *cfg.Config, cdc *codec.Codec) error {
 			return err
 		}
 
-		// Gather gentxs folder
+		// gather gentxs folder
 		err = writeFile(fmt.Sprintf("%v.json", nodeDirName), gentxsDir, txBytes)
 		if err != nil {
 			_ = os.RemoveAll(outDir)
 			return err
 		}
 	}
 
-	// Generate empty genesis.json
+	if err := initGenFiles(cdc, chainID, accs, genFiles, numValidators); err != nil {
+		return err
+	}
+
+	err := collectGenFiles(
+		cdc, config, chainID, monikers, nodeIDs, valPubKeys, numValidators,
+		outDir, viper.GetString(flagNodeDirPrefix), viper.GetString(flagNodeDaemonHome),
+	)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Successfully initialized %d node directories\n", numValidators)
+	return nil
+}
+
+func initGenFiles(
+	cdc *codec.Codec, chainID string, accs []app.GenesisAccount,
+	genFiles []string, numValidators int,
+) error {
+
 	appGenState := app.NewDefaultGenesisState()
 	appGenState.Accounts = accs
+
 	appGenStateJSON, err := codec.MarshalJSONIndent(cdc, appGenState)
 	if err != nil {
 		return err
 	}
+
 	genDoc := types.GenesisDoc{
 		ChainID:    chainID,
 		AppState:   appGenStateJSON,
 		Validators: nil,
 	}
 
-	// Save all genesis.json files
+	// generate empty genesis files for each validator and save
 	for i := 0; i < numValidators; i++ {
 		if err := genDoc.SaveAs(genFiles[i]); err != nil {
 			return err
 		}
 	}
 
+	return nil
+}
+
+func collectGenFiles(
+	cdc *codec.Codec, config *cfg.Config, chainID string,
+	monikers, nodeIDs []string, valPubKeys []crypto.PubKey,
+	numValidators int, outDir, nodeDirPrefix, nodeDaemonHomeName string,
+) error {
+
+	var appState json.RawMessage
+	genTime := tmtime.Now()
+
 	for i := 0; i < numValidators; i++ {
-		nodeDirName := fmt.Sprintf("%s%d", viper.GetString(nodeDirPrefix), i)
-		nodeDaemonHomeName := viper.GetString(nodeDaemonHome)
+		nodeDirName := fmt.Sprintf("%s%d", nodeDirPrefix, i)
 		nodeDir := filepath.Join(outDir, nodeDirName, nodeDaemonHomeName)
 		gentxsDir := filepath.Join(outDir, "gentxs")
 		moniker := monikers[i]
 		config.Moniker = nodeDirName
+
 		config.SetRoot(nodeDir)
 
 		nodeID, valPubKey := nodeIDs[i], valPubKeys[i]
-		// Run `init` and generate genesis.json and config.toml
 		initCfg := initConfig{
 			ChainID:   chainID,
 			GenTxsDir: gentxsDir,
 			Name:      moniker,
 			NodeID:    nodeID,
 			ValPubKey: valPubKey,
 		}
+
 		genDoc, err := loadGenesisDoc(cdc, config.GenesisFile())
 		if err != nil {
 			return err
 		}
-		if _, err := genAppStateFromConfig(cdc, config, initCfg, genDoc); err != nil {
+
+		nodeAppState, err := genAppStateFromConfig(cdc, config, initCfg, genDoc)
+		if err != nil {
+			return err
+		}
+
+		if appState == nil {
+			// set the canonical application state (they should not differ)
+			appState = nodeAppState
```

Comment:
It's weird that we're using a "canonical" state here, and yet... we call genAppStateFromConfig 4 times.... and it's also weird that we "initGenFiles" which writes mostly files, then load them again...

I think we should be creating a single genesis doc once, and calling WriteGenesisFile(genFile, chainID, nil, appState, genTime) 4 times instead.  Basically, most of this logic doesn't belong in for-loop.

"collectGenFiles" is also a misnomer.  While it's collecting gentxs, this function does not "collect" gen files.

Basically, we should not have 2 steps of "init" and "collect"... we should init the gendoc once and "write" 4 times.

Reply:
Yes @jaekwon. The original logic was not great to begin with. It initially was one giant function. I simply broke the steps out to auxiliary functions. I believe that was the right approach. However, I agree that we must clean this up drastically and I'll take care of that.